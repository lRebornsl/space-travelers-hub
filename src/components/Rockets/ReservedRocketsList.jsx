import { useSelector } from 'react-redux';

const ReservedRocketsList = () => {
  const rockets = useSelector((state) => state.rocketsSlice.rockets);
  const reservedRockets = rockets.filter((rocket) => rocket.reserved);

  return (
    <div>
      <h2>My Rockets</h2>
      {reservedRockets.length === 0 ? (
        <p>No rockets reserved yet.</p>
      ) : (
        <ul>
          {reservedRockets.map((rocket) => (
            <li key={rocket.id}>
              {rocket.name}
            </li>
          ))}
        </ul>
      )}
    </div>
  )
}

export default ReservedRocketsList
