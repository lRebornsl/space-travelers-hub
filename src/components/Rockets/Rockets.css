.rockets-container {
  padding: 1rem 2rem;
}

.rocket-card-container {
  display: flex;
  gap: 1rem;
}

.rocket-img {
  width: 22rem;
}

.rocket-name {
  font-weight: 600;
}

.rocket-description {
  font-size: 1.4rem;
  color: #484747;
  font-weight: 500;
  margin-top: 10px;
}

.reserve-btn-container {
  margin-top: 1rem;
}
