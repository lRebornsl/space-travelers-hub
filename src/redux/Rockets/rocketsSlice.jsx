import { createAsyncThunk, createReducer } from '@reduxjs/toolkit';
import { URL } from '../../utils/constant';

const FETCH_ROCKETS = 'rocketStore/rockets/FETCH_ROCKETS';
const RESERVE_ROCKET = 'rocketStore/rockets/RESERVE_ROCKETS';

export const fetchRockets = createAsyncThunk(FETCH_ROCKETS, async (post, { dispatch }) => {
  const response = await fetch(URL);
  const data = await response.json();
  const payload = data.map((rocket) => ({
    id: rocket.rocket_id,
    name: rocket.rocket_name,
    description: rocket.description,
    type: rocket.rocket_type,
    image: rocket.flickr_images[0],
    reserved: false,
  }));
  dispatch({
    type: FETCH_ROCKETS,
    payload,
  });
  return payload;
})

export const reserveRocket = (rocketId) => ({
  type: RESERVE_ROCKET,
  payload: rocketId,
});

const initialState = {
  rockets: [],
}

const rocketsSlice = createReducer(initialState, (builder) => {
  builder
    .addCase(fetchRockets.fulfilled, (state, { payload }) => ({
    ...state,
    rockets: [...payload],
    }))
    .addCase(RESERVE_ROCKET, (state, { payload: rocketId }) => ({
      ...state,
      rockets: state.rockets.map((rocket) =>
        rocket.id !== rocketId ? rocket : { ...rocket, reserved: true }
      ),
    }));
})

export default rocketsSlice;