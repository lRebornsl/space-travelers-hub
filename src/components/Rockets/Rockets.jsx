import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveRocket, cancelReservation } from '../../redux/Rockets/rocketsSlice';
import Button from '../Button/Button';
import './Rockets.css';

const Rockets = ({ rocket }) => {
  const dispatch = useDispatch();
  
  const handleReserveClick = () => {
    if (rocket.reserved) {
      dispatch(cancelReservation(rocket.id))
    } else {
      dispatch(reserveRocket(rocket.id))
    }
  }

  return (
    <div className="rockets-container">
      <div className="rocket-card-container">
        <div>
          <img src={rocket.image} alt="rocket image" className="rocket-img" />
        </div>
        <div className="rocket-info-container">
          <div>
            <p className="rocket-name">{rocket.name}</p>
          </div>
          <div>
            <p className="rocket-description">{rocket.description}</p>
          </div>
          <div className="reserve-btn-container">
            {rocket.reserved ? (
              <Button
                name="Cancel Reservation"
                onclick={handleReserveClick}
                stylename="cancel-reserve-btn"
              />
            ) : (
              <Button
                name="Reserve Rocket"
                onclick={handleReserveClick}
                stylename="reserve-btn"
              />
            ) }
          </div>
        </div>
      </div>
    </div>
  )
}

Rockets.propTypes = {
  rocket: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    reserved: PropTypes.bool.isRequired,
  }).isRequired,
  
}

export default Rockets;

