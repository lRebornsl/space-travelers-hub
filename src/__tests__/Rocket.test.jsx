import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import Rockets from '../components/Rockets/Rockets';
import { describe, expect, test } from 'vitest';
import { Provider } from 'react-redux';
import store from '../redux/Store';
import { MemoryRouter } from 'react-router-dom';

describe('Badge component', () => {
  const mockRocket = {
    name: 'Falcon 9',
    description: 'A two-stage orbital launch vehicle.',
    reserved: false,
  };

  test('renders rocket name and description correctly', () => {
    const { getByText } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Rockets rocket={mockRocket} />
        </Provider>
      </MemoryRouter>
    );
    const nameElement = getByText(mockRocket.name);
    const descriptionElement = getByText(mockRocket.description);
    expect(nameElement).toBeInTheDocument();
    expect(descriptionElement).toBeInTheDocument();
  });

  test('renders "Reserved Rocket" button if rocket is not reserved', () => {
    const { getByText } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Rockets rocket={mockRocket} />
        </Provider>
      </MemoryRouter>
    );
    const reserveButton = getByText('Reserve Rocket');
    expect(reserveButton).toBeInTheDocument();
  })
})