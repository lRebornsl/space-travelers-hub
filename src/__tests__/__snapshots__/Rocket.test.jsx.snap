// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Badge component > Rocket Page renders correctly into the DOM 1`] = `
<section
  className="container"
/>
`;
